docs/rationale.md

```diff
@@ -23,4 +23,69 @@ try (Scope scope = TracingContextUtils.currentContextWith(span)) {
 ```
 
 It would not be possible to call `recordException` if `span` was also using try-with-resources.
-Because this is a common usage for spans, we do not support try-with-resources.
\ No newline at end of file
+Because this is a common usage for spans, we do not support try-with-resources.
+
+
+## Versioning and Releases
+
+### Assumptions
+
+- This project uses semver v2, as does the rest of OpenTelemetry.
+
+### Goals
+
+- API Stability: 
+    - Once the API for a given signal (spans, logs, metrics, baggage) has been officially released, that API module will 
+function, *with no recompilation required*, with any SDK that has the same major version, and equal or greater minor or patch version. 
+    - For example, libraries that are instrumented with `opentelemetry-api-trace:1.0.1` will function with 
+SDK library `opentelemetry-sdk-trace:1.11.33`. 
+    - We call this requirement the "ABI" compatibility requirement for "Application Binary Interface" compatibility. 
+- SDK Stability:
+    - Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. 
+    - Internal interfaces are allowed to break. 
+    - Strict ABI compatibility is not required.
```

Comment:
This can't be allowed right? I think many SDK classes are still too important (`OpenTelemetrySdk.Builder` for example, span processors, sampler) to risk dependency hell by breaking ABI. Imagine a custom sampler someone published, we can't break it.

Reply:
I think this is worth having a discussion about. I'm pretty hesitant to say we're going to maintain ABI compatibility across the entire swath of the SDK at this point, given how youthful it is. A custom sampler should be something that it is easy for an SDK user to update, correct? It's not going to be backed into a library or instrumentation, is it?

Reply:
It requires the sampler to also update to the newer SDK first, something we shouldn't force post-1.0 for SDK either. For whatever reason, it might be slow to update, or may be abandoned but still useful - the moment we create a situation where a user can't upgrade OTel because they have a dependency linked to the SDK, which IMO is quite valid), then we lose most of the value of semver in the first place. If we didn't, I'd say we'd need to list even trace as `sdk-trace-experimental`, we need the experimental flag to reflect whether we are going to introduce dependency hell on users or not.

Another example is Spring, which will be using the SDK classes. If our SDK is 1.0, we should be in a state where we don't cause version conflicts between Spring and OTel (Spring 4.2.3 requires OTel SDK 1.3.4, yikes).

Reply:
So... what SDK APIs are going to be locked in stone forever? How do we distinguish one public SDK API from another? I feel like if we're going to require strict ABI compatibility with some subset of the SDK, we need a way to publish that module separately, as something like `opentelemetry-sdk-api` or something. Any ideas on how to achieve this goal?

Reply:
All of it I think - the reason we have any public methods at all is because we expect code outside our control to use them. As soon as that happens, we need to commit to those public methods, or we'll cause conflicts - any library that causes conflicts becomes unpopular fast. I don't see anything special about the SDK vs the API in this regard.

We should be as diligent as we can by 1.0 to get things into a good place - we probably won't get everything and we'll have methods deprecated until the next major, etc. It is what it is, but it's safer than introducing ABI breaks that could cause dependency locks for users.

Reply:
First, I agree with Anuraag, that SDK should have stable public interface. With the same API/ABI compatibility guarantees as API.

Second, as I read it, OTEP 143 allows as to stabilise SDK separate from API. Thus we don't have to claim SDK stability yet. We can just release stable API and continue working on SDK.

Third, if we are going to provide stability guarantees for SDK, we have to distinguish public and internal parts of it. And guarantee only public parts. How to achieve that: separate packages, separate artifacts, JPMS, OSGI - is a separate question :)

Reply:
yes, after consideration and discussion, I also agree with this. I don't know how we're going to achieve it, but I also think it's important. 

Ideas on how to effectively version the various parts of the project independently are very welcome!

Reply:
Ya, being able to release breaking SDK 2.0 without having to bump API to 2.0 is nice

I think we have similar issue in instrumentation repo, e.g. being able to release grpc library instrumentation 2.0 without having to bump `opentelemetry-instrumentation-api`.

This will require a bunch more release infra to manage versioning/releasing modules independently (and using prefixed tags, e.g. `api-1.1`, `sdk-2.0`).